### DISCUSSION TITLE
How to know if user has touched the form with controller component?

### DISCUSSION THREAD

**Discussion by hk-skit** [react-hook-form/react-hook-form#2638] [Q&A] [answered]
I have forked a CSB found form API documentation to figure out the dirty and touched state of the form. The `isDirty` flag from the `formState` resets to `false` if user inputs the initial/default values. The `touched` object doesn't seem to update with the controller/custom component. How do we figure if the form is touched or not? Ideally `isDirty` shouldn't reset to false.

Codesandbox: https://codesandbox.io/s/react-hook-form-formstate-dirty-touched-submitted-forked-z3ihi

**Comment by bluebill1049** [comment, marked as answer]
> Ideally `isDirty` shouldn't reset to false.

what do you mean? if the user change the form into the initial state then it's not dirty.

**Reply by hk-skit** [reply]
Thanks for the speedy reply. I believe that a form is dirty if a user has interacted with it. If that's not the correct definition of dirty then how do I figure out if a user has interacted/touched the form. In CSB, when the user interacts with the custom component, `touched` form state doesn't change. Is there any way to figure it out?

**Reply by bluebill1049** [reply]
nope, we follow standard here. perhaps you are after `touched`.

**Reply by hk-skit** [reply]
I used `touched` but it doesn't contain form fields rendered with the `Controller`. Is there any way to add "controller" fields to `touched` object when the user interacts with it? Maybe something like `setTouched` which can be invoked in `onchange` handler

```
const MyInput = ({ onChange, onBlur, value, name }) => (
  <input
    onBlur={onBlur}
    onChange={(e) => onChange(e.target.value)}
    value={value}
    name={name}
  />
);
```

**Reply by bluebill1049** [reply]
> I used `touched` but it doesn't contain form fields rendered with the `Controller`.

provide a codesandbox plz.

**Reply by hk-skit** [reply]
I have added a `MyCounter` component in the same sandbox. It doesn't get added to the `touched` state.

CSB: https://codesandbox.io/s/react-hook-form-formstate-dirty-touched-submitted-forked-z3ihi?file=/src/index.js

**Reply by bluebill1049** [reply]
I don't see `touched`.

**Reply by bluebill1049** [reply]
<img width="1034" alt="Screen Shot 2020-08-18 at 8 09 00 pm" src="https://user-images.githubusercontent.com/10513364/90500756-a9e23000-e18e-11ea-870b-c096bf9971b4.png">

**Reply by hk-skit** [reply]
> I don't see touched.

Sorry I didn't get this part. Is there anything specific I need to do in `MyCounter` component?

```
const MyCounter = ({ onChange, value }) => {
  return (
    <div style={{ display: `flex` }}>
      <button type="button" onClick={() => onChange(`${+value + 1}`)}>
        +
      </button>
      <span style={{ color: `white` }}>{value}</span>
      <button type="button" onClick={() => onChange(`${+value - 1}`)}>
        -
      </button>
    </div>
  );
};
```

**Reply by bluebill1049** [reply]
you should fire up `onBlur`, to trigger `touched`

**Reply by hk-skit** [reply]
Ah! I see. That's why `MyInput` is not having this issue. Thank you very much @bluebill1049. I appreciate your help.

**Reply by bluebill1049** [reply]
no worries :) glad helped.
